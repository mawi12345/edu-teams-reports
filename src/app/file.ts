import csv from 'csv-parser';
import { Row, FileContent, Student } from './types';

export function csvBlobToTable(file: Blob): Promise<Row[]> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onabort = () => reject(new Error('file reading was aborted'));
    reader.onerror = () => reject(new Error('file reading has failed'));
    reader.onload = () => {
      const parser = csv({ separator: ',', headers: false });
      const results: object[] = [];
      parser.on('data', data => results.push(data));
      parser.on('end', () => {
        resolve(results as Row[]);
      });
      parser.write(reader.result);
      parser.end();
    };
    reader.readAsText(file);
  });
}

export function isFileSupported(rows: Row[]): boolean {
  const headerRow = rows[0];
  return (
    headerRow[0] === 'First Name' &&
    headerRow[1] === 'Last Name' &&
    headerRow[2] === 'Email Address'
  );
}

const saveNumber = (a: number) => {
  if (a && !isNaN(a)) {
    return a;
  }
  return 0;
};

export function parseFile(rows: Row[]): FileContent {
  const content: FileContent = {
    students: [],
    todos: [],
    sum: 0,
  };
  const headerRow = Object.values(rows[0]) as string[];
  const count = Math.round((headerRow.length - 3) / 3);

  for (let x = 0; x < count; x += 1) {
    content.todos.push({
      name: headerRow[3 + x * 3],
      max: 0,
    });
  }

  for (let i = 1; i < rows.length; i += 1) {
    const studentRow = Object.values(rows[i]) as string[];

    if (i === 1) {
      for (let x = 0; x < count; x += 1) {
        const max = parseInt(studentRow[4 + x * 3], 10);
        if (max && !isNaN(max)) {
          content.todos[x].max = max;
          content.sum += max;
        }
      }
    }

    const results: (number | undefined)[] = [];
    let max = 0;
    for (let x = 0; x < count; x += 1) {
      if (studentRow[3 + x * 3] === '') {
        console.log(`${studentRow[1]} fehlt bei ${content.todos[x].name}`);
        results.push(undefined);
      } else {
        max += content.todos[x].max;
        results.push(saveNumber(parseInt(studentRow[3 + x * 3], 10)));
      }
    }

    const studentSum: number = results.reduce<number>(
      (a: number | undefined, b: number | undefined) => (a || 0) + (b || 0),
      0,
    );

    content.students.push({
      first: studentRow[0],
      last: studentRow[1],
      sum: studentSum,
      max,
      sumPercent: Math.round((studentSum * 100) / max),
      results,
    });
  }

  return content;
}

export const sortContent = (orig: FileContent): FileContent => {
  const todos = orig.todos.map((todo, index) => {
    const results = orig.students.map(student => student.results[index]);
    return {
      ...todo,
      results,
    };
  });
  todos.sort((a, b) => {
    const numA = parseInt(a.name);
    const numB = parseInt(b.name);

    let ac: number | string = a.name.toUpperCase();
    let bc: number | string = b.name.toUpperCase();
    if (!isNaN(numA) && !isNaN(numB)) {
      ac = numA;
      bc = numB;
    }

    if (ac < bc) {
      return -1;
    }
    if (ac > bc) {
      return 1;
    }
    return 0;
  });

  const students: Student[] = orig.students.map((student, index) => ({
    ...student,
    results: todos.map(t => t.results[index]),
  }));

  students.sort((a, b) => {
    const nameA = a.last.toUpperCase();
    const nameB = b.last.toUpperCase();
    if (nameA < nameB) {
      return -1;
    }
    if (nameA > nameB) {
      return 1;
    }
    return 0;
  });

  return {
    students,
    sum: orig.sum,
    todos,
  };
};
