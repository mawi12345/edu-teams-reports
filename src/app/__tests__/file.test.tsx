import { csvBlobToTable, parseFile } from '../file';

const testData = `"Vorname","Nachname","E-Mail-Adresse","<name-of-task1>","Punkt","Feedback","<name-of-task2>","Punkt","Feedback","<name-of-task3>","Punkt","Feedback"
"Vorname1","Nachname1","1@schule.at","6","7","","20","25","","","10",""
"Vorname2","Nachname2","2@schule.at","5","7","","20","25","","9","10",""
"Vorname3","Nachname3","3@schule.at","7","7","","20","25","","10","10",""
`;

describe('file', () => {
  it('should convert blob to table', async () => {
    const blob = new Blob([testData], {
      type: 'text/plain',
    });
    const row = await csvBlobToTable(blob);
    expect(row).toMatchSnapshot();
  });

  it('should parse table', async () => {
    const blob = new Blob([testData], {
      type: 'text/plain',
    });
    const row = await csvBlobToTable(blob);
    const content = parseFile(row);
    expect(content).toMatchSnapshot();
  });
});
