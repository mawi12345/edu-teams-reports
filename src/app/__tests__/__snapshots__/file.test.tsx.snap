// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file should convert blob to table 1`] = `
Array [
  Object {
    "0": "Vorname",
    "1": "Nachname",
    "10": "Punkt",
    "11": "Feedback",
    "2": "E-Mail-Adresse",
    "3": "<name-of-task1>",
    "4": "Punkt",
    "5": "Feedback",
    "6": "<name-of-task2>",
    "7": "Punkt",
    "8": "Feedback",
    "9": "<name-of-task3>",
  },
  Object {
    "0": "Vorname1",
    "1": "Nachname1",
    "10": "10",
    "11": "",
    "2": "1@schule.at",
    "3": "6",
    "4": "7",
    "5": "",
    "6": "20",
    "7": "25",
    "8": "",
    "9": "",
  },
  Object {
    "0": "Vorname2",
    "1": "Nachname2",
    "10": "10",
    "11": "",
    "2": "2@schule.at",
    "3": "5",
    "4": "7",
    "5": "",
    "6": "20",
    "7": "25",
    "8": "",
    "9": "9",
  },
  Object {
    "0": "Vorname3",
    "1": "Nachname3",
    "10": "10",
    "11": "",
    "2": "3@schule.at",
    "3": "7",
    "4": "7",
    "5": "",
    "6": "20",
    "7": "25",
    "8": "",
    "9": "10",
  },
]
`;

exports[`file should parse table 1`] = `
Object {
  "students": Array [
    Object {
      "first": "Vorname1",
      "last": "Nachname1",
      "max": 32,
      "results": Array [
        6,
        20,
        undefined,
      ],
      "sum": 26,
      "sumPercent": 81,
    },
    Object {
      "first": "Vorname2",
      "last": "Nachname2",
      "max": 42,
      "results": Array [
        5,
        20,
        9,
      ],
      "sum": 34,
      "sumPercent": 81,
    },
    Object {
      "first": "Vorname3",
      "last": "Nachname3",
      "max": 42,
      "results": Array [
        7,
        20,
        10,
      ],
      "sum": 37,
      "sumPercent": 88,
    },
  ],
  "sum": 42,
  "todos": Array [
    Object {
      "max": 7,
      "name": "<name-of-task1>",
    },
    Object {
      "max": 25,
      "name": "<name-of-task2>",
    },
    Object {
      "max": 10,
      "name": "<name-of-task3>",
    },
  ],
}
`;
